openff/toolkit/topology/molecule.py

```diff
@@ -5785,11 +5789,42 @@ def sort_hierarchy_elements(self):
         their identifiers.
         """
 
-        # hard-code the sort_func value here, since it's hard to serialize safely
-        def sort_func(x):
-            return version.parse(".".join([str(i) for i in x.identifier]))
+        def _sort_chains(element: HierarchyScheme) -> str:
+            return str(element.identifier[0])
+
+        def _sort_residues(element: HierarchyScheme) -> tuple[str, int]:
+            return (
+                str(element.identifier[0]),
+                int(element.identifier[1]),
+                str(element.identifier[2]),
+                str(element.identifier[3]),
+            )
+
+        def _sort_other(element: HierarchyScheme) -> tuple[str, int]:
+            x = list()
+            for identifier in element.identifier:
+                # If an identifier is int-ish, cast it to int,
+                # otherwise let it remain as a string
+                try:
+                    x.append(int(identifier))
+                except ValueError:
+                    x.append(identifier)
+
+            return tuple(x)
+
```

Comment:
This is inelegant but the best I can come up with at the moment given the number of constraints

* One has no idea what the length of `HierarchyElement.identifier` is a prior
* It is implicit but not required that elements should be sorted by identifiers in the order they're provided at construction
  * For chains, this is unimportant given there is only one identifier
  * For residues, this shouldn't be important unless a single residue number in a chain maps to multiple insertion codes and/or residue names (?)
* Residue numbers can be stored as string representations of numbers
* Empty values can sneak in as `" "` and `"None"` but never `None`
* 100% of the time, our users care about chains and residues but some effort must be made to sort other arbitrary hierarchy schemes